### PR TITLE
implemented emulate prepare and added tests

### PR DESCRIPTION
**Pull request by v-kaywon**
Fix for #92 and #140 .

Implement emulate prepare by:
1. When a statement is prepared with emulate prepare, parse out placeholders and store it in a placeholders hashtable in stmt
2. Get statement information in the quote function
3. Going through the placeholders in order, get the encoding from stmt->bound_params
4. Set the encoding to param level encoding, stmt level encoding, or connection level encoding; param encoding dominate, then stmt level, then connection level. If one level's encoding is invalid, then go to the higher level encoding.
5. "quote" the parameter base on the encoding
6. Replace placeholders in the sql string with proper quoted/encoded parameters

Note: Emulate prepare only works for input bound parameters.
When user specify PDO::PARAM_INPUT_OUTPUT in PDO::bindParam when emulate prepare is on, the driver throws a PDO exception.
When user binds an output parameter, the sql query should be a placeholder preceding an '=' sign (e.g., "SELECT **? =** COUNT(*) FROM Employee WHERE Department='Human Resources'"). When while parsing out the placeholders out from the sql string, it also checks if the sql query has an '=' immediately after a placeholder. If it does, throws an exception saying output parameter with emulate prepared statement is not supported. If user binds an output parameter to a stored procedure, an error would not be raise.

### CODE REVIEW

**Comment by david-garcia-garcia** [comment]
Thaaaaanks!! Just an question... What happens with duplicate placeholders in the updated implementation? On the previous driver you could not have a placeholder twice, yet on other PDO drivers such as MySQL they will let you re-use parameteres by allowing a placeholder to be repeated.

**Comment by v-kaywon** [comment]
@david-garcia-garcia ,

Can you clarify on what you mean by duplicated placeholders?

Do you mean having two placeholders with the same name in a sql query?
For example.
```
$statement = $connection->prepare("SELECT * FROM Table1 WHERE uid=:id AND pid=:id");
$statement->bindParam(':id', 1234);
```
And both ':id' placeholders in the sql query is replaced with 1234.

Or do you mean binding to the same placeholder twice?
For example,
```
$statement = $connection->prepare("SELECT * FROM Table1 WHERE uid=:id");
$statement->bindParam(':id', 1234);
$statement->bindParam(':id', 6789);
```

**Comment by david-garcia-garcia** [comment]
I'm talking about your first example werwe the placeholder is present twice in the query itself.

**Comment by v-kaywon** [comment]
I did some preliminary testing. Seems like duplicated placeholders works for parameters of any encoding types that's not binary.

File: source/pdo_sqlsrv/pdo_dbh.cpp [hunk 1]
```diff
@@ -1279,13 +1289,67 @@ int pdo_sqlsrv_dbh_quote( pdo_dbh_t* dbh, const char* unquoted, size_t unquoted_
     PDO_VALIDATE_CONN;
     PDO_LOG_DBH_ENTRY;
 
-    pdo_sqlsrv_dbh* driver_dbh = reinterpret_cast<pdo_sqlsrv_dbh*>( dbh->driver_data );
-    SQLSRV_ENCODING encoding = driver_dbh->bind_param_encoding;
+    SQLSRV_ENCODING encoding = SQLSRV_ENCODING_CHAR;
+	
+    // get the current object in PHP; this distinguishes pdo_sqlsrv_dbh_quote being called from:
+	// 1. PDO::quote() - object name is PDO
+	// 2. PDO::execute() - object name is PDOStatement
```
Comment (Hadis-Fard) [inline source/pdo_sqlsrv/pdo_dbh.cpp]: I think you meant PDOStatement::execute 😉

**Pull request merged**